### PR TITLE
[spi_host] Test more spi host functionality 

### PR DESCRIPTION
**Pull request by cfrantz**
Add spi_host tests that exercise more features of the spi_host.  This
test now depends on several bits of EEPROM functionality; fortunately,
the IS25WP016 part on the CW310 board supports all of the operations.

1. Test single opcode transactions with no data phase (software_reset, chip_erase).
2. Test read transactions with dummy cycles (read_sfdp, fast_read).
3. Test read and write transactions without dummy cycles (page_program).
4. Test two-lane reads (dual_read).
5. Test four-lane reads (quad_read).

Note: this PR depends on #17615 and #17617.  You need only review the last 3 commits in this PR.

### CODE REVIEW

**Review by andreaskurth — COMMENTED** [review]
Great that you're adding a test with different opcodes + reads/writes between `spi_host` and a real EEPROM, thanks @cfrantz! I've only reviewed the last commit (the actual test), which LGTM.

One question: Could we extend this test to include reads/writes with `addr_is_4b = true` (and a sufficiently wide address)?

CC @hcallahan-lowrisc and @johngt for visibility

**Comment by cfrantz** [comment]
> One question: Could we extend this test to include reads/writes with `addr_is_4b = true` (and a sufficiently wide address)?

We can, but not with the IS25WP016 chip that is currently on the CW310 board.  That part doesn't appear to support any form of 4-byte addressing.

**Comment by hcallahan-lowrisc** [comment]
@cfrantz Would you prefer to swap out this flash for another part? It could certainly be considered if it would be helpful.
(I'm looking at the SPI/I2C breakout boards for Host testing, so may be ordering some hardware in the near future and I could consider a new part for this socket as well.)

**Comment by cfrantz** [comment]
> @cfrantz Would you prefer to swap out this flash for another part? It could certainly be considered if it would be helpful. (I'm looking at the SPI/I2C breakout boards for Host testing, so may be ordering some hardware in the near future and I could consider a new part for this socket as well.)

Yes, I would.  But I think this test should be submitted as-is and enhanced once we have the new parts installed in CI and distributed to developers.

**Comment by hcallahan-lowrisc** [comment]
> > @cfrantz Would you prefer to swap out this flash for another part? It could certainly be considered if it would be helpful. (I'm looking at the SPI/I2C breakout boards for Host testing, so may be ordering some hardware in the near future and I could consider a new part for this socket as well.)
> 
> Yes, I would. But I think this test should be submitted as-is and enhanced once we have the new parts installed in CI and distributed to developers.

Yes, sorry for the tangent, I will track elsewhere. Thanks

**Comment by johngt** [comment]
Thanks for the cc @andreaskurth . Also tagging @engdoreis as he will be going through SPI test plan and needs to be aligned.

**Review by a-will — APPROVED** [review]

File: sw/device/lib/testing/spi_flash_testutils.c [hunk 1]
```diff
@@ -85,29 +86,65 @@ void spi_flash_testutils_read_sfdp(dif_spi_host_t *spih, uint32_t address,
                                         ARRAYSIZE(transaction)));
 }
 
+status_t spi_flash_testutils_read_status(dif_spi_host_t *spih, uint8_t opcode,
+                                         size_t length) {
+  CHECK(spih != NULL);
```
Comment (engdoreis) [inline sw/device/lib/testing/spi_flash_testutils.c:91]: Maybe `TRY_CHECK`?

File: sw/device/lib/testing/spi_flash_testutils.c [hunk 2]
```diff
@@ -85,29 +86,65 @@ void spi_flash_testutils_read_sfdp(dif_spi_host_t *spih, uint32_t address,
...
+      },
+  };
+  TRY(dif_spi_host_transaction(spih, /*csid=*/0, transaction,
+                               ARRAYSIZE(transaction)));
+  return OK_STATUS(status);
+}
+
+status_t spi_flash_testutils_write_status(dif_spi_host_t *spih, uint8_t opcode,
+                                          uint32_t status, size_t length) {
+  CHECK(spih != NULL);
```
Comment (engdoreis) [inline sw/device/lib/testing/spi_flash_testutils.c:116]: `TRY_CHECK`

File: sw/device/lib/testing/spi_flash_testutils.c [hunk 3]
```diff
@@ -205,3 +248,142 @@ void spi_flash_testutils_program_page(dif_spi_host_t *spih, uint8_t *payload,
...
+      // Device does not have a QE bit.
+      break;
+    case 1:
+      // QE is bit1 of status reg 2.
+      // Set/clear via two-byte reads/writes via SR1 opcodes.
+      // Writing only one byte to SR1 clears SR2.
+      status = TRY(spi_flash_testutils_read_status(
+          spih, kSpiDeviceFlashOpReadStatus1, 2));
+      status = bitfield_bit32_write(status, 9, enabled);
+      status = TRY(spi_flash_testutils_write_status(
```
Comment (engdoreis) [inline sw/device/lib/testing/spi_flash_testutils.c:324]: `status` is not used later, so maybe:
```suggestion
    TRY(spi_flash_testutils_write_status(
```

File: sw/device/lib/testing/spi_flash_testutils.h [hunk 4]
```diff
@@ -41,6 +42,36 @@ typedef enum spi_flash_status_bit {
   kSpiFlashStatusBitWel = 0x2,
 } spi_flash_status_bit_t;
 
+/**
+ * Perform a Read Status command.
+ *
+ * Issues a Read Status transaction using the requested opcode.
+ * In the case of a multi-byte status, the bytes are assembled and returned
+ * as a litte-endian word
```
Comment (engdoreis) [inline sw/device/lib/testing/spi_flash_testutils.h:77]: Nit.
```suggestion
 * as a litte-endian word.
```

File: sw/device/lib/testing/spi_flash_testutils.h [hunk 5]
```diff
@@ -41,6 +42,36 @@ typedef enum spi_flash_status_bit {
...
+ */
+status_t spi_flash_testutils_read_status(dif_spi_host_t *spih, uint8_t opcode,
+                                         size_t length);
+
+/**
+ * Perform a Write Status command.
+ *
+ * Issues a Write Status transaction using the requested opcode.
+ * In the case of a multi-byte status, the status word bytes are
+ * as a litte-endian word
```
Comment (engdoreis) [inline sw/device/lib/testing/spi_flash_testutils.h:92]: Nit.
```suggestion
 * as a litte-endian word.
```

File: sw/device/lib/testing/spi_flash_testutils.h [hunk 6]
```diff
@@ -63,22 +94,61 @@ void spi_flash_testutils_issue_write_enable(dif_spi_host_t *spih);
...
+ *
+ * @param spih A SPI host handle.
+ * @param opcode The desired program opcode.
+ * @param payload A pointer to the payload to be written to the page.
+ * @param length Number of bytes in the payload. Must be less than or equal to
+ *               256 bytes.
+ * @param address The start address where the payload programming should begin.
+ *                Note that an address + length that crosses a page boundary may
+ *                wrap around to the start of the page.
+ * @param addr_is_4b True if `address` is 4 bytes long, else 3 bytes.
```
Comment (engdoreis) [inline sw/device/lib/testing/spi_flash_testutils.h:169]: Why not use `dif_spi_host_addr_mode_t` here?

File: sw/device/lib/testing/spi_flash_testutils.h [hunk 7]
```diff
@@ -96,4 +166,35 @@ void spi_flash_testutils_program_page(dif_spi_host_t *spih, uint8_t *payload,
...
+                                 uint8_t *payload, size_t length,
+                                 uint32_t address, bool addr_is_4b,
+                                 uint8_t width, uint8_t dummy);
+
+/**
+ * Enable or disable Quad mode on the EEPROM according to the SFDP-described
+ * method.
+ *
+ * @param spih A SPI host handle.
+ * @param method The method to use to enable.  This value should come from the
```
Comment (engdoreis) [inline sw/device/lib/testing/spi_flash_testutils.h:218]: Nit.
```suggestion
 * @param method The method to use to enable. This value should come from the
```

File: sw/device/tests/spi_host_smoketest.c [hunk 8]
```diff
@@ -18,54 +21,136 @@
...
+
+  op.opcode = kSpiDeviceFlashOpResetEnable;
+  TRY(dif_spi_host_transaction(&spi_host, /*cs_id=*/0, &op, 1));
+  return OK_STATUS();
+}
+
+status_t test_read_sfdp(void) {
+  spi_flash_testutils_read_sfdp(&spi_host, 0, sfdp.data, sizeof(sfdp.data));
+  LOG_INFO("SFDP signature is 0x%08x", sfdp.header.signature);
+  CHECK(sfdp.header.signature == kSfdpSignature,
```
Comment (engdoreis) [inline sw/device/tests/spi_host_smoketest.c:64]: `TRY_CHECK`?

**Review by engdoreis — COMMENTED** [review]
Thanks @cfrantz  for the PR, I didn't notice that we had a spi_flash_testutils, should this be refactored to only return status_t as well? If so, I can do it after this PR is merged.

File: sw/device/lib/testing/spi_flash_testutils.c (hunk 1 above)
Comment (cfrantz) [reply sw/device/lib/testing/spi_flash_testutils.c:91]: Lets defer that for a follow-on PR that cleans up spi_flash_testutils with respect to `status_t`.

File: sw/device/lib/testing/spi_flash_testutils.c (hunk 2 above)
Comment (cfrantz) [reply sw/device/lib/testing/spi_flash_testutils.c:116]: Lets defer that for a follow-on PR that cleans up spi_flash_testutils with respect to status_t.

File: sw/device/tests/spi_host_smoketest.c (hunk 8 above)
Comment (cfrantz) [reply sw/device/tests/spi_host_smoketest.c:64]: The test program needs to abort if the SFDP signature is incorrect -- no other meaningful test can be performed if the flash fails to identify properly.

File: sw/device/lib/testing/spi_flash_testutils.h (hunk 7 above)
Comment (cfrantz) [reply sw/device/lib/testing/spi_flash_testutils.h:218]: https://www.thesaurus.com/e/writing/is-there-1-space-or-2-after-a-period/
"""
According to research, the preference for two spaces after a period falls along generational lines. Studies have shown that, beginning with millennials, younger generations widely prefer the single space after a period. On the flip side, Gen X and older generations usually prefer the double space.
"""

Am Gen-X.  I claim the exemption.

**Pull request merged**